File: src/homework05/mvp/Presenter.java
package homework05.mvp;

import homework05.exceptions.DuplicateContactException;
import homework05.exceptions.DuplicatePhoneException;
import homework05.exceptions.InvalidPhoneException;
import homework05.exceptions.PhoneNotFoundException;
import homework05.services.FileFormat;

import java.util.Map;

public class Presenter {

    private final Model model;
    private final View view;

    public Presenter(View view, Map<FileFormat, String> paths) {
        this.view = view;
        model = new Model(paths);
    }

    public void loadFromFile(FileFormat fileFormat) {
        model.load(fileFormat);

        if (model.currentBook().size() > 0) {
            view.showMessage("Phone book loaded successfully");
        }
    }

    public boolean addContact() {
        try {
            model.addContact(view.getStringFieldFromInput("first name"), view.getStringFieldFromInput("last name"));
        } catch (DuplicateContactException e) {
            view.showMessage(e.getMessage());
            return false;
        }
        view.showMessage("Contact added successfully");
        return true;
    }

    public void removeContact() {
        model.currentBook().remove(model.currentContact());
        view.showMessage("Contact removed successfully");

        if (model.currentBook().size() > 0) {
            model.getContact(0);
        } else {
            model.setCurrentContact(null);
            view.showMessage("Phone book is empty");
        }
    }

    public void addPhone() {
        try {
            model.currentContact().addPhone(view.getStringFieldFromInput("phone number"));
            view.showMessage("Phone added successfully");
        } catch (InvalidPhoneException | DuplicatePhoneException e) {
            view.showMessage(e.getMessage());
        }
    }

    public void removePhone() {
        try {
            model.currentContact().removePhone(view.getStringFieldFromInput("phone number"));
            view.showMessage("Phone removed successfully");
        } catch (InvalidPhoneException | PhoneNotFoundException e) {
            view.showMessage(e.getMessage());
        }
    }

    public boolean getContact() {
        int id = getId();
        if (id < 0) {
            view.showMessage("Wrong value");
            return false;
        }
        model.getContact(id);
        return true;
    }

    private int getId() {
        view.inputPrompt("contact ID");
        return view.getIntegerFieldFromInput();
    }

    public boolean findContacts() {
        String name = view.getStringFieldFromInput("search string");
        int id;
        int numberOfContacts = model.findContacts(name);
        if (numberOfContacts > 1) {
            id = contactChoiceMenu();
        } else if (numberOfContacts == 1) {
            id = 0;
        } else {
            view.showMessage("Contacts not found");
            return false;
        }

        if (id >= 0) {
            setCurrentContactFromFound(0);
            return true;
        }
        view.showMessage("Wrong value");
        return false;
    }

    private int contactChoiceMenu() {
        view.showMessage(model.showFoundContacts());
        return getId();
    }

    public void saveToFile(FileFormat fileFormat) {
        String msg;
        if (model.save(fileFormat)) msg = "File successfully saved";
        else {
            msg = "Failed to save file";
        }
        view.showMessage(msg);
    }

    public void showPhonebook() {
        view.showMessage(model.showPhonebook());
    }

    public void setCurrentContactFromFound(int index) {
        model.setCurrentContact(model.foundContacts().get(index));
    }

    public void showCurrentContact() {
        view.showMessage(model.showCurrentContact());
    }

    public void newPhonebook() {
        model.newPhonebook();
    }

}

File: src/homework07/phonebook/mvp/presenter.py
from phonebook.exceptions.phonebook_exceptions import DuplicateContactException, DuplicatePhoneException, \
    InvalidPhoneException, PhoneNotFoundException
from phonebook.mvp.model import Model


class Presenter:
    def __init__(self, view, paths) -> None:
        self._view = view
        self._model = Model(paths)

    def add_contact(self):
        try:
            self._model.add_contact(self._view.get_string_field_from_input('first name'),
                                    self._view.get_string_field_from_input('last name'))
        except DuplicateContactException as ex:
            self._view.show_message(ex)
            return False
        self._view.show_message('Contact added successfully')
        return True

    def remove_contact(self):
        try:
            self._model.current_book.remove(self._model.current_contact)
            self._view.show_message('Contact removed successfully')
        except (KeyError, IndexError) as ex:
            self._view.show_message('Failed to remove contact')
            self._view.show_message(ex)

    def add_phone(self):
        try:
            self._model.current_contact.add_phone(self._view.get_string_field_from_input('phone number'))
            self._view.show_message('Phone added successfully')
        except (InvalidPhoneException, DuplicatePhoneException) as ex:
            self._view.show_message(ex)

    def remove_phone(self):
        try:
            self._model.current_contact.remove_phone(self._view.get_string_field_from_input('phone number'))
            self._view.show_message('Phone removed successfully')
        except (InvalidPhoneException, PhoneNotFoundException) as ex:
            self._view.show_message(ex)

    def get_contact(self):
        index = self.get_id()
        if index < 0:
            self._view.show_message('Wrong value')
            return False
        try:
            self._model.get_contact(index)
        except IndexError as ex:
            self._view.show_message(ex)
            return False
        return True

    def get_id(self):
        self._view.input_prompt('contact ID')
        return self._view.get_int_from_input()

    def find_contacts(self):
        name = self._view.get_string_field_from_input('search string')
        number_of_contacts = self._model.find_contacts(name)
        if number_of_contacts > 1:
            index = self.contact_choice_menu()
        elif number_of_contacts == 1:
            index = 0
        else:
            self._view.show_message('Contacts not found')
            return False

        if 0 <= index < number_of_contacts:
            self.set_current_contact_from_found(index)
            return True

        self._view.show_message('Wrong value')
        return False

    def contact_choice_menu(self):
        self._view.show_message(self._model.show_found_contacts())
        return self.get_id()

    def set_current_contact_from_found(self, index):
        self._model.current_contact = self._model.found_contacts[index]

    def show_phonebook(self):
        self._view.show_string(self._model.show_phonebook())

    def show_current_contact(self):
        self._view.show_string(self._model.show_current_contact())

    def new_phonebook(self):
        self._model.new_phonebook()

    def load_from_file(self, file_format):
        if self._model.load(file_format):
            msg = 'Phone book loaded successfully'
        else:
            msg = 'Failed to load phone book'
        self._view.show_message(msg)

    def save_to_file(self, file_format):
        if self._model.save(file_format):
            msg = 'File successfully saved'
        else:
            msg = 'Failed to save file'
        self._view.show_message(msg)
